packages/composable-controller/src/ComposableController.ts

```diff
@@ -157,27 +160,29 @@ export class ComposableController<
 
     super({
       name: controllerName,
-      metadata: controllers.reduce<StateMetadata<ComposableControllerState>>(
-        (metadata, controller) => ({
-          ...metadata,
-          [controller.name]: isBaseController(controller)
-            ? controller.metadata
-            : { persist: true, anonymous: true },
-        }),
-        {} as never,
-      ),
-      state: controllers.reduce<ComposableControllerState>(
+      metadata: Object.keys(controllers).reduce<
+        StateMetadata<ComposableControllerState>
+      >((metadata, name) => {
+        (metadata as StateMetadataConstraint)[name] = {
+          persist: true,
+          anonymous: true,
+        };
+        return metadata;
+      }, {} as never),
+      state: Object.values(controllers).reduce<ComposableControllerState>(
         (state, controller) => {
-          return { ...state, [controller.name]: controller.state };
+          (state as ComposableControllerStateConstraint)[controller.name] =
+            controller.state;
+          return state;
```

Comment:
Similar to above, can we avoid the type assertion here by doing:

``` suggestion
          return {
            ...state,
            [controller.name]: controller.state,
          };
```

Reply:
This is not desirable, as apparently reduce operations that don't reuse their output object run four orders of magnitude slower than ones that do.
> https://consensys.slack.com/archives/C07KB8HRZ4J/p1732211272904539
> http://measurethat.net/Benchmarks/Show/21554/0/objectfromentries-vs-reduce-vs-assign
> https://github.com/MetaMask/metamask-mobile/pull/12374

The type assertion is unfortunate, but it doesn't seem like a critical compromise in type safety, given that the generic type will be resolved in downstream environments, even if the end result is not apparent to TypeScript here. 

Also, in this case, the potential performance gains seemed like a higher priority.

Reply:
Applied here: https://github.com/MetaMask/core/pull/4968/commits/61c96974cee9447cf91fd7b16775b47749e894ec